frontend/request.go

```diff
@@ -0,0 +1,118 @@
+	"github.com/moby/buildkit/frontend/dockerfile/parser"
+	"github.com/moby/buildkit/frontend/dockerui"
+	gwclient "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/moby/buildkit/solver/pb"
+	"github.com/pkg/errors"
+)
+
+type solveRequestOpt func(*gwclient.SolveRequest) error
+
+func newSolveRequest(opts ...solveRequestOpt) (gwclient.SolveRequest, error) {
+	var sr gwclient.SolveRequest
+
+	for _, o := range opts {
+		if err := o(&sr); err != nil {
+			return sr, err
+		}
+	}
+	return sr, nil
+}
+
+func toFrontend(f *dalec.Frontend) solveRequestOpt {
+	return func(req *gwclient.SolveRequest) error {
+		req.Frontend = gatewayFrontend
+		if req.FrontendOpt == nil {
+			req.FrontendOpt = make(map[string]string)
+		}
+		req.FrontendOpt["source"] = f.Image
+		req.FrontendOpt["cmdline"] = f.CmdLine
+		return nil
+	}
+}
+
+func withSpec(ctx context.Context, spec *dalec.Spec, opts ...llb.ConstraintsOpt) solveRequestOpt {
+	return func(req *gwclient.SolveRequest) error {
+		if req.FrontendInputs == nil {
+			req.FrontendInputs = make(map[string]*pb.Definition)
+		}
+
+		dt, err := yaml.Marshal(spec)
+		if err != nil {
+			return errors.Wrap(err, "error marshalling spec to yaml")
+		}
+
+		def, err := llb.Scratch().File(llb.Mkfile(dockerui.DefaultDockerfileName, 0600, dt), opts...).Marshal(ctx)
+		if err != nil {
+			return errors.Wrap(err, "error marshaling spec to LLB")
+		}
+		req.FrontendInputs[dockerui.DefaultLocalNameDockerfile] = def.ToPB()
+		return nil
+	}
+}
+
+func withTarget(t string) solveRequestOpt {
+	return func(req *gwclient.SolveRequest) error {
+		if req.FrontendOpt == nil {
+			req.FrontendOpt = make(map[string]string)
+		}
+		req.FrontendOpt["target"] = t
+		return nil
+	}
+}
+
+func toDockerfile(ctx context.Context, bctx llb.State, dt []byte, spec *dalec.SourceBuild, opts ...llb.ConstraintsOpt) solveRequestOpt {
+	return func(req *gwclient.SolveRequest) error {
+		req.Frontend = "dockerfile.v0"
+
+		bctxDef, err := bctx.Marshal(ctx)
+		if err != nil {
+			return errors.Wrap(err, "error marshaling dockerfile to LLB")
+		}
+		if req.FrontendInputs == nil {
+			req.FrontendInputs = make(map[string]*pb.Definition)
+		}
+
+		dfDef, err := marshalDockerfile(ctx, dt, opts...)
+		if err != nil {
+			return errors.Wrap(err, "error marshaling dockerfile to LLB")
+		}
+
+		req.FrontendInputs[dockerui.DefaultLocalNameContext] = bctxDef.ToPB()
+		req.FrontendInputs[dockerui.DefaultLocalNameDockerfile] = dfDef.ToPB()
+
+		if ref, cmdline, _, ok := parser.DetectSyntax(dt); ok {
+			req.Frontend = gatewayFrontend
+			if req.FrontendOpt == nil {
+				req.FrontendOpt = make(map[string]string)
+			}
+			req.FrontendOpt["source"] = ref
+			req.FrontendOpt["cmdline"] = cmdline
+		}
+
+		if spec != nil {
+			if req.FrontendOpt == nil {
+				req.FrontendOpt = make(map[string]string)
+			}
+			if spec.Target != "" {
+				req.FrontendOpt["target"] = spec.Target
+			}
+			for k, v := range spec.Args {
+				req.FrontendOpt["build-arg:"+k] = v
```

Comment:
is this `req.FrontendOpts` pointing to the build-args pass when calling `docker buildx build --build-arg "<key-vaule>"`? If so shouldn't this have greater precedence against the ones set on the spec `arg` block?
If that is the case we would need to do a conditional to prevent overwrite?

Reply:
At this point we've already loaded all the build args, including filling in the values from the client into the spec.
Additionally, these are args set on a build source, e.g.

```yaml
sources:
  someBuild:
    build:
      args:
        someKey: someval
```